Python/Product/VSInterpreters/CondaEnvironmentFactoryProvider.cs

```diff
@@ -0,0 +1,372 @@
+    [Export(typeof(CondaEnvironmentFactoryProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    class CondaEnvironmentFactoryProvider : IPythonInterpreterFactoryProvider {
+        private readonly Dictionary<string, PythonInterpreterInformation> _factories = new Dictionary<string, PythonInterpreterInformation>();
+        internal const string FactoryProviderName = "CondaEnv";
+        internal const string EnvironmentCompanyName = "CondaEnv";
+        private int _ignoreNotifications;
+        private bool _initialized;
+        private readonly CPythonInterpreterFactoryProvider _globalProvider;
+        private readonly bool _watchFileSystem;
+        private FileSystemWatcher _envsTxtWatcher;
+        private Timer _envsTxtWatcherTimer;
+        private string _environmentsTxtPath;
+
+        [ImportingConstructor]
+        public CondaEnvironmentFactoryProvider(
+            [Import] CPythonInterpreterFactoryProvider globalProvider,
+            [Import("Microsoft.VisualStudioTools.MockVsTests.IsMockVs", AllowDefault = true)] object isMockVs = null
+        ) {
+            _watchFileSystem = isMockVs == null;
+            _globalProvider = globalProvider;
+        }
+
+        private void EnsureInitialized() {
+            lock (this) {
+                if (!_initialized) {
+                    _initialized = true;
+                    _environmentsTxtPath = Path.Combine(
+                        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
+                        ".conda",
+                        "environments.txt"
+                    );
+
+                    DiscoverInterpreterFactories();
+
+                    if (_watchFileSystem) {
+                        // Watch the file %HOMEPATH%/.conda/Environments.txt which
+                        // is updated by conda after a new environment is created.
+                        var watchedPath = Path.GetDirectoryName(_environmentsTxtPath);
+                        if (Directory.Exists(watchedPath)) {
+                            try {
+                                _envsTxtWatcher = new FileSystemWatcher(watchedPath, "*.txt");
+                                _envsTxtWatcher.Changed += _envsTxtWatcher_Changed;
+                                _envsTxtWatcher.Created += _envsTxtWatcher_Changed;
+                                _envsTxtWatcher.EnableRaisingEvents = true;
+                                _envsTxtWatcherTimer = new Timer(_envsTxtWatcherTimer_Elapsed);
+                            } catch (ArgumentException) {
+                            } catch (IOException) {
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private async void _envsTxtWatcherTimer_Elapsed(object state) {
+            try {
+                _envsTxtWatcherTimer.Change(Timeout.Infinite, Timeout.Infinite);
+            } catch (ObjectDisposedException) {
+            }
```

Comment:
If we're disposed, do we still want to discover?

Reply:
That was probably an unnecessary precaution given that nobody disposes of the Timer (the factory provider is not disposable). If it was to become disposable, I agree it makes no sense to discover. I'll move it inside the try block.